synapse/util/caches/response_cache.py

```diff
@@ -112,6 +117,31 @@ def remove(r):
         result.addBoth(remove)
         return result.observe()
 
+    def add_conditional(self, key: T, conditional: Callable[[Any], bool]):
+        self.pending_conditionals.setdefault(key, set()).add(conditional)
```

Comment:
this seems like a dangerous thing to add to the public interface. It seems like it would be very easy to use in racy way.

Reply:
should i signal that `add_conditional` is private by prefixing it with `_` (in a new PR)?

Reply:
since it's only called in one place, I'd just inline it.